python/tk_multi_workfiles/entity_models/extended_model.py

```diff
@@ -0,0 +1,280 @@
+
+        Leaves are only considered if the given Entity type matches the Entity
+        type this model represents. Otherwise, the full model hierarchy is traversed
+        to retrieve the given Entity.
+
+        .. note::
+            The same entity can appear multiple times in the hierarchy, the first
+            match is returned.
+
+        :param str entity_type: A Shotgun Entity type.
+        :param int entity_id: The Shotgun id of the Entity to look for.
+        """
+        # If dealing with the primary entity type this model represents, just
+        # call the base implementation which only considers leaves.
+        if entity_type == self.get_entity_type():
+            return super(ShotgunExtendedEntityModel, self).item_from_entity(
+                entity_type, entity_id
+            )
+        # If not dealing with the primary entity type, we need to traverse the
+        # model to find the entity.
+        # Bail out quickly if we know that the entity type we are looking for is
+        # not in this model.
+        # Please note that this implies that we need to load all intermediate nodes
+        # in the tree to get an accurate list of all the entity types it contains.
+        if entity_type not in self._entity_types:
+            return None
+        # If the model is empty, just bail out
+        if not self.rowCount():
+            return None
+        parent_list = [self.invisibleRootItem()]
+        while parent_list:
+            parent = parent_list.pop()
+            for row_i in range(parent.rowCount()):
+                item = parent.child(row_i)
+                if self.canFetchMore(item.index()):
+                    self.fetchMore(item.index())
+                entity = self.get_entity(item)
+                # If dealing with an item holding the entity type we're looking
+                # for, we don't add it to the list of items to explore. We return
+                # it if the id is the one we're looking for.
+                if not entity or entity["type"] != entity_type:
+                    if item.hasChildren():
+                        parent_list.append(item)
+                elif entity["id"] == entity_id:
+                    return item
+        return None
+
+    def item_from_field_value_path(self, field_value_list):
+        """
+        Retrieve an item from a list of field values identifying its path.
+
+        This allows to retrieve an item in an updated model from a list of
+        collected field values representing its path.
+
+        Full or partial matches are performed: if the item represented by the
+        given value list is not present in the model anymore, the last item
+        matched from the value list is returned.
+
+        :param field_value_list: A list of field values for the path from the
+                                 root to the item.
+        """
+        if not self.rowCount():
+            return None
+        parent = self.invisibleRootItem()
```

Comment:
Not sure I understand your comment: we have a `fetchMore` so we dynamically load in the model items which were cached?

Reply:
I'm thinking of something like
Episode->Sequences->Shot->Step->Task. Let's say Shot's children haven't been fetched from Shotgun yet. canFetchMore will return False, but you won't be able to retrieve Step or Task in the next iteration of the loop because it is asynchronously being loaded.

Reply:
This is the extended model, not the deferred one, so all data is already pulled from SG, but the model is may be not yet populated with them. I think I added a comment on the deferred model explaining that new data is not fetched from SG if not already pulled.